### PR TITLE
Add from_ieee754_32 Presto function

### LINKED ISSUE #4427: implement from_ieee754_32 function for presto
### Description

Please assign it to me, thanks!

### PR DESCRIPTION
**Pull request by pdabre12**
Fixes #4427 

Reference: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java#L267

### CODE REVIEW

File: velox/functions/prestosql/tests/BinaryFunctionsTest.cpp [hunk 1]
```diff
@@ -697,4 +697,38 @@ TEST_F(BinaryFunctionsTest, toIEEE754Bits32) {
...
+      fromIEEE754Bits32(hexToDec("7f800000")));
+  EXPECT_EQ(
+      -std::numeric_limits<float>::infinity(),
+      fromIEEE754Bits32(hexToDec("ff800000")));
+  EXPECT_THROW(fromIEEE754Bits32("YQ"), VeloxUserError);
+  EXPECT_EQ(3.4028235E38f, fromIEEE754Bits32(toIEEE754Bits32(3.4028235E38f)));
+  EXPECT_EQ(-3.4028235E38f, fromIEEE754Bits32(toIEEE754Bits32(-3.4028235E38f)));
+  EXPECT_EQ(1.4E-45f, fromIEEE754Bits32(toIEEE754Bits32(1.4E-45f)));
+  EXPECT_EQ(-1.4E-45f, fromIEEE754Bits32(toIEEE754Bits32(-1.4E-45f)));
+  EXPECT_EQ(
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/BinaryFunctionsTest.cpp:729]: Can you do the same for std::numeric_limit<float>::nan(), quiet_Nan(), signaling_Nan() as well ?

Also add tests for numeric_limits::min() and max()

**Review by aditi-pandit — COMMENTED** [review]
Thanks @pdabre12.

Code looks good. Small comment in the tests.

**Comment by aditi-pandit** [comment]
@pdabre12 : Please can you add results of ExpressionFuzzer with this function https://facebookincubator.github.io/velox/develop/testing/fuzzer.html#how-to-run

**Comment by pdabre12** [comment]
This test fails because 

Expected equality of these values:
```
std::numeric_limits<float>::quiet_NaN()
    Which is: nan
  fromIEEE754Bits32( toIEEE754Bits32(std::numeric_limits<float>::quiet_NaN()))
  Which is: (nan)

```

Any suggestions of what I can do about the parenthesis () surrounding nan?

**Comment by czentgr** [comment]
> This test fails because
> 
> Expected equality of these values:
> 
> ```
> std::numeric_limits<float>::quiet_NaN()
>     Which is: nan
>   fromIEEE754Bits32( toIEEE754Bits32(std::numeric_limits<float>::quiet_NaN()))
>   Which is: (nan)
> ```
> 
> Any suggestions of what I can do about the parenthesis () surrounding nan?

As discussed. You can't compare using `EXPECT_EQ` because this is the same as nan == nan which always returns false ("A NaN never compares equal to itself." from the cpp reference). Instead, check that the return from the function call is actually a nan via `std::isnan` and `EXPECT_TRUE`.

**Review by aditi-pandit — APPROVED** [review]
Thanks @pdabre12 . Code is good.

**Review by aditi-pandit — CHANGES_REQUESTED** [review]
@pdabre12 : The linux-build has test failures. Please fix the tests as per Christian's comments.

**Comment by pdabre12** [comment]
@aditi-pandit Please review. The pipeline is clean now.

**Comment by pdabre12** [comment]
Fuzzer output:

```
I20240205 18:16:46.861337 22035276 ExpressionVerifier.cpp:153] Starting simplified eval execution.
I20240205 18:16:46.861505 22035276 FuzzerToolkit.cpp:103] Exceptions match.
I20240205 18:16:46.861514 22035276 ExpressionFuzzerVerifier.cpp:363] ==============================> Done with iteration 39674
I20240205 18:16:46.861541 22035276 ExpressionFuzzerVerifier.cpp:160] ==============================> Top 1 by number of rows processed
I20240205 18:16:46.861547 22035276 ExpressionFuzzerVerifier.cpp:162] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20240205 18:16:46.861552 22035276 ExpressionFuzzerVerifier.cpp:166] from_ieee754_32 39675 100.00% 2896347
I20240205 18:16:46.861565 22035276 ExpressionFuzzerVerifier.cpp:172] ==============================> Bottom 1 by number of rows processed
I20240205 18:16:46.861572 22035276 ExpressionFuzzerVerifier.cpp:174] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20240205 18:16:46.861577 22035276 ExpressionFuzzerVerifier.cpp:179] from_ieee754_32 39675 100.00% 2896347
I20240205 18:16:46.861582 22035276 ExpressionFuzzerVerifier.cpp:192] ==============================> All stats sorted by number of times the function was chosen
I20240205 18:16:46.861588 22035276 ExpressionFuzzerVerifier.cpp:194] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20240205 18:16:46.861593 22035276 ExpressionFuzzerVerifier.cpp:198] from_ieee754_32 39675 100.00% 2896347
E20240205 18:16:46.861600 22035276 ExpressionFuzzerVerifier.cpp:369] Total iterations: 39675
E20240205 18:16:46.861605 22035276 ExpressionFuzzerVerifier.cpp:370] Total failed: 32138
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

**Review by aditi-pandit — APPROVED** [review]
Thanks

**Comment by aditi-pandit** [comment]
@Yuhta , @kagamiori : Please can you help with the review and merge.

**Comment by ashokku2022** [comment]
@Yuhta @kagamiori I know you are busy with other high-priority issues, however, I was wondering if you could look into this one whenever you get time so that we can close it. Thanks!

File: velox/docs/functions/presto/binary.rst [hunk 2]
```diff
@@ -94,6 +94,10 @@ Binary Functions
 
     Encodes ``real`` in a 32-bit big-endian binary according to IEEE 754 single-precision floating-point format.
 
+.. function:: from_ieee754_32(binary) -> real
```
Comment (aditi-pandit) [inline velox/docs/functions/presto/binary.rst:97]: Please move this to the right alphabetical order in the list.

**Review by aditi-pandit — COMMENTED** [review]
@pdabre12 : Had a quick review comment.

File: velox/functions/prestosql/BinaryFunctions.h [hunk 3]
```diff
@@ -424,4 +424,18 @@ struct ToIEEE754Bits32 {
...
 
+template <typename T>
+struct FromIEEE754Bits32 {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<float>& result,
+      const arg_type<Varbinary>& input) {
+    static constexpr auto kTypeLength = sizeof(int32_t);
+    VELOX_USER_CHECK_EQ(input.size(), kTypeLength, "Expected 4-byte input");
```
Comment (aditi-pandit) [inline velox/functions/prestosql/BinaryFunctions.h:435]: The java error is "Input floating-point value must be exactly 4 bytes long"

File: velox/functions/prestosql/tests/BinaryFunctionsTest.cpp [hunk 4]
```diff
@@ -697,4 +697,58 @@ TEST_F(BinaryFunctionsTest, toIEEE754Bits32) {
...
+  const auto fromIEEE754Bits32 = [&](const std::optional<std::string>& arg) {
+    return evaluateOnce<float, std::string>(
+        "from_ieee754_32(c0)", {arg}, {VARBINARY()});
+  };
+
+  const auto toIEEE754Bits32 = [&](std::optional<float> arg) {
+    return evaluateOnce<std::string, float>("to_ieee754_32(c0)", arg);
+  };
+
+  EXPECT_EQ(std::nullopt, fromIEEE754Bits32(std::nullopt));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/BinaryFunctionsTest.cpp:711]: Add a test case where you check the invalid input > 4 bytes case.

**Review by aditi-pandit — CHANGES_REQUESTED** [review]
Please see comments above.

**Comment by pdabre12** [comment]
@aditi-pandit Addressed the comments, please review again. 
Thanks.

**Review by aditi-pandit — APPROVED** [review]
Thanks @pdabre12

**Comment by aditi-pandit** [comment]
@Yuhta @kagamiori : Please can you help with review and merge.

File: velox/docs/functions/presto/binary.rst [hunk 5]
```diff
@@ -26,6 +26,14 @@ Binary Functions
 
     Decodes binary data from the hex encoded ``string``.
 
+.. function:: from_ieee754_32(binary) -> real
+
+    Decodes the 32-bit big-endian ``binary`` in IEEE 754 single-precision floating-point format.
```
Comment (Yuhta) [inline velox/docs/functions/presto/binary.rst:31]: Would it be possible to clarify what happens if input is shorter / longer than 32 bits?

Comment (pdabre12) [reply velox/docs/functions/presto/binary.rst:31]: @Yuhta Thanks for the review.
Added a line in the docs. PTAL.

Comment (Yuhta) [reply velox/docs/functions/presto/binary.rst:31]: I think it actually throws user error (`VeloxUserError`)

Comment (pdabre12) [reply velox/docs/functions/presto/binary.rst:31]: You are right, it does throw a `VeloxUserError`.
I tried to keep it generic as `VeloxUserError` is also a `VeloxException`, but changed it to `VeloxUserError`.
Please take another look.

**Pull request closed without merging**